zkml/src/quantization/strategy.rs

```diff
@@ -217,105 +153,103 @@ impl InferenceTracker {
-                        let (quant_min_output, _quant_max_output) =
-                            quantized_conv.output_range(*quantization::MIN, *quantization::MAX);
-                        // TODO: remove this is broken
-                        let output_scaling = ScalingFactor::default();
-                        md.set_layers_scaling(id, output_scaling);
-                        let shift =
-                            last_input_scaling_factor.shift(&model_scaling, &output_scaling);
-                        last_input_scaling_factor = output_scaling;
-                        let requant = Requant::new(quant_min_output.abs() as usize, shift);
-                        vec![Layer::Convolution(quantized_conv), Layer::Requant(requant)]
-                    }
-                    a => {
-                        return vec![a.quantize(
-                            &last_input_scaling_factor,
-                            None, // no scaling factor for bias needed for this layer
-                        )];
-                    }
-                }
-            })
-            .collect::<Vec<Layer<Element>>>();
-        Ok((
-            Model::<Element>::new_from(quantized_layers, input_not_padded_shape, input_shape),
-            md.build(),
-        ))
+        quantize_model::<AbsoluteMax>(model, (), input_scaling_factor)
+    }
+}
+
+fn quantize_model<S: ScalingStrategy>(
+    model: Model<f32>,
+    data: S::AuxData,
+    input_scaling: Vec<ScalingFactor>,
+) -> anyhow::Result<(Model<Element>, ModelMetadata)>
+where
+    Dense<f32>: QuantizeOp<S, QuantizedOp = Dense<Element>>,
+    Convolution<f32>: QuantizeOp<S, QuantizedOp = Convolution<Element>>,
+{
+    let input_shapes = model.input_shapes();
+    let input_not_padded_shapes = model.unpadded_input_shapes();
+    let mut md = MetadataBuilder::new(input_scaling);
+    // 2. Create the requant layers from the infered data
+    let mut requant_layers = vec![];
+    let mut catch_err: Result<()> = Ok(());
+    let nodes = model
+        .into_forward_iterator()
+        .map(|(node_id, node)| {
+            let input_scaling = md.compute_input_scaling(&node.inputs)?;
+            let quantized_out =
+                quantize_op::<S, _>(node.operation, &data, node_id, &input_scaling)?;
+            md.set_layers_scaling(node_id, quantized_out.output_scalings, input_scaling);
+            if let Some(requant) = quantized_out.requant_layer {
+                requant_layers.push((node_id, requant));
+            }
+            let quantized_node =
+                Node::new_with_outputs(node.inputs, quantized_out.quanzited_op, node.outputs);
+            Ok((node_id, quantized_node))
+        })
+        .map_while(|n| {
+            if n.is_err() {
+                catch_err = Err(n.unwrap_err());
+                None
+            } else {
+                Some(n.unwrap())
+            }
+        });
+    let mut model = Model::new_from_shapes(input_not_padded_shapes, input_shapes, nodes);
```

Comment:
I understand the concerns about not collecting first while keeping the iterator argument on the model creation... but failing early is often, if not always, a better choice. 
Would it be acceptable to give a hashmap to Model ? i dont see it as an horror show since anyway it's just another way to represent the same information `(node_id, node)` which is imo even cleaner. 
Said differently, we can keep the same memory consumption at the cost of changing the container of the info `(node_id, node)` and yet being able to panic early in case of problems.

Reply:
Done in commit https://github.com/Lagrange-Labs/deep-prove/pull/114/commits/fa37beb0683993623225ff7b66e3fbfa99564baf